src/measure_theory/lp_space.lean

```diff
@@ -0,0 +1,126 @@
+
+## Main definitions
+
+* `mem_ℒp f p μ` : the function `f` has finite p-seminorm for measure `μ`, for `p:ℝ` such that
+                  `hp1 : 1 ≤ p`,
+
+## Notation
+
+* `snorm f p μ` : `(∫⁻ a, (nnnorm (f a))^(p : ℝ) ∂μ) ^ (1/p)` for `f : α → β`
+
+-/
+
+open measure_theory
+
+noncomputable theory
+
+namespace ℒp_space
+
+variables {α β γ: Type*} [measurable_space α] {μ : measure α}
+  [measurable_space β] [normed_group β]
+  [normed_group γ]
+  {p : ℝ}
+
+section ℒp_space_definition
+
+/-- The property that f belongs to ℒp for measure μ and real p -/
+def mem_ℒp (f : α → β) (p : ℝ) (μ : measure α) : Prop :=
+measurable f ∧ ∫⁻ a, (nnnorm (f a)) ^ p ∂μ < ⊤
+
+/-- seminorm on ℒp -/
+def snorm (f : α → γ) (p : ℝ) (μ : measure α) : ennreal := (∫⁻ a, (nnnorm (f a))^p ∂μ) ^ (1/p)
+
+lemma lintegral_rpow_nnnorm_eq_rpow_snorm {f : α → γ} (hp0_lt : 0 < p) :
+  ∫⁻ a, (nnnorm (f a)) ^ p ∂μ = (snorm f p μ) ^ p :=
+begin
+  unfold snorm,
+  rw [←ennreal.rpow_mul, one_div, inv_mul_cancel, ennreal.rpow_one],
+  symmetry, exact ne_of_lt hp0_lt,
+end
+
+end ℒp_space_definition
+
+lemma mem_ℒp_one_iff_integrable : ∀ f : α → β, mem_ℒp f 1 μ ↔ integrable f μ :=
+begin
+  intro f,
+  unfold integrable, unfold has_finite_integral, unfold mem_ℒp,
+  simp only [ennreal.rpow_one, nnreal.coe_one],
+end
+
+section top
+
+lemma snorm_lt_top {f : α → β} (hp0 : 0 ≤ p) (hfp : mem_ℒp f p μ) : snorm f p μ < ⊤ :=
+begin
+  unfold snorm,
+  refine ennreal.rpow_lt_top_of_nonneg _ (ne_of_lt hfp.right),
+  rw [one_div, inv_nonneg],
+  exact hp0,
+end
+
+lemma snorm_ne_top {f : α → β} (hp0 : 0 ≤ p) (hfp : mem_ℒp f p μ) : snorm f p μ ≠ ⊤ :=
+ne_of_lt (snorm_lt_top hp0 hfp)
+
+lemma lintegral_rpow_nnnorm_lt_top_of_snorm_lt_top {f : α → γ} (hp0_lt : 0 < p)
+  (hfp : snorm f p μ < ⊤) :
+  ∫⁻ a, (nnnorm (f a)) ^ p ∂μ < ⊤ :=
+begin
+  rw lintegral_rpow_nnnorm_eq_rpow_snorm hp0_lt,
+  exact ennreal.rpow_lt_top_of_nonneg (le_of_lt hp0_lt) (ne_of_lt hfp),
+end
+
+lemma mem_ℒp_of_snorm_lt_top {f : α → β} (hp0_lt : 0 < p) (hfm : measurable f)
+  (hfp : snorm f p μ < ⊤) :
+  mem_ℒp f p μ :=
+begin
+  split,
+  exact hfm,
+  exact lintegral_rpow_nnnorm_lt_top_of_snorm_lt_top hp0_lt hfp,
+end
+
+end top
+
+section zero
+
+lemma lintegral_rpow_nnnorm_zero (hp0_lt : 0 < p) : ∫⁻ a, (nnnorm ((0 : α → γ) a))^p ∂μ = 0 :=
+begin
+  simp_rw pi.zero_apply,
+  rw [nnnorm_zero, lintegral_const, mul_eq_zero],
+  left,
+  exact ennreal.zero_rpow_of_pos hp0_lt,
+end
+
+lemma zero_mem_ℒp (hp0_lt : 0 < p): mem_ℒp (0 : α → β) p μ :=
+begin
+  split,
+  exact measurable_zero,
+  exact lt_of_le_of_lt (le_of_eq (lintegral_rpow_nnnorm_zero hp0_lt)) with_top.zero_lt_top,
+end
+
+lemma snorm_zero (hp0_lt : 0 < p): snorm (0 : α → γ) p μ = 0 :=
+begin
+  have h : ∫⁻ a, (nnnorm ((0 : α → γ) a))^(p : ℝ) ∂μ = 0,
+  from lintegral_rpow_nnnorm_zero hp0_lt,
+  unfold snorm,
+  rw [h, ennreal.rpow_eq_zero_iff],
+  left,
+  split,
+  refl,
+  rw [one_div, inv_pos], exact hp0_lt,
```

Comment:
Do you know that `simp` is extremely useful and powerful? Here, these 4 lines (that should have been 5 as the last line should be split in two) can be replaced simply by `simp [hp0_lt]`. But even better, the whole proof can be replaced with `by simp [snorm, hp0_lt]` (try it!). This is the result of a careful design of the library, in which enough lemmas that "should" be applied whenever they can be applied, are marked with the `simp` attribute. By the way, I think the current `snorm_zero` would deserve being marked as such a simp lemma.

Reply:
I know about simp but I do not have the right feel for when it can help me and when it cannot. I try it a lot in places where it does absolutely nothing useful, and apparently miss some places where it can produce a one-line proof :)

Are there simple criteria for lemmas that should be @[simp] ?

Reply:
There is no universal critertion to know what to mark as `@[simp]`. Roughly, if a lemma transforms an expression into a simpler one, and you feel it should be applied essentially whenever you see the left-hand side, to turn into the right-hand side, then this is a good simp lemma.